common/src/tor.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright 2021. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::{io, net::TcpListener};
+
+use libtor::{LogDestination, LogLevel, TorFlag};
+use log::*;
+use multiaddr::Multiaddr;
+use rand::{distributions::Alphanumeric, thread_rng, Rng};
+use tari_shutdown::ShutdownSignal;
+use tempfile::{tempdir, NamedTempFile, TempDir, TempPath};
+use tor_hash_passwd::EncryptedKey;
+
+use crate::{exit_codes::ExitCodes, CommsTransport, TorControlAuthentication};
+
+const LOG_TARGET: &str = "common::tor";
+
+pub struct Tor {
+    control_port: u16,
+    data_dir: String,
+    log_destination: String,
+    log_level: LogLevel,
+    passphrase: Option<String>,
+    socks_port: u16,
+    temp_dir: Option<TempDir>,
+    temp_file: Option<TempPath>,
+}
+
+impl Default for Tor {
+    fn default() -> Tor {
+        Tor {
+            control_port: 19_051,
+            data_dir: "/tmp/tor-data".into(),
+            log_destination: "/tmp/tor.log".into(),
+            log_level: LogLevel::Err,
+            passphrase: None,
+            socks_port: 19_050,
+            temp_dir: None,
+            temp_file: None,
+        }
+    }
+}
+
+impl Tor {
+    /// Returns a new Tor instance with random options.
+    /// The data directory, passphrase, and log destination are temporary and randomized.
+    /// Two available adjacent TCP ports will be selected in the given
```

Comment:
Think this comment needs to be updated. RE my comment on your previous PR, Ah I realise that you cannot give tor the 0 port and have to bind yourself because you need to know the port that the OS chooses. Let's see if it ever becomes an issue with already being bound then unbound / OS gives the port to some other process. Seems like that would be rare at worst. Otherwise, I suppose a configured port is the way to go.

Reply:
Good catch, thanks will update the comment. You'll see a bit lower down I do actually bind to `0` and then use `local_addr` to retrieve the port number

Reply:
Yeah, but you still have to bind on the port yourself, then release it - giving the tor control port a 0 will allocate a port, but you don't have an easy way to get which port was chosen unless you could ask tor through the control port, but that is a mission.I think that on most OS's it should unbind the port in time for tor to rebind on it, I seem to remember having this problem long ago but maybe things have changed / windows problem 🤷  - I'm sure it'll be 💯